### PR TITLE
[IMP] mail, Add mark as read action to message

### PR DESCRIPTION
**Pull request by maki-odoo**
By this commit, user can mark a message as read and moves it from inbox to history mailbox

### CODE REVIEW

File: addons/mail/static/src/new/core/messaging.js [hunk 1]
```diff
@@ -372,18 +373,14 @@ export class Messaging {
...
-                            message.needaction_partner_ids.splice(partnerIndex, 1);
-                        }
-                        const messageIndex = this.state.discuss.inbox.messages.find(
-                            (m) => m.id === message.id
-                        );
-                        if (messageIndex !== -1) {
-                            this.state.discuss.inbox.messages.splice(messageIndex, 1);
+                        removeFromArray(message.needaction_partner_ids, partnerIndex);
+                        removeFromArray(this.state.discuss.inbox.messages, messageId);
+                        console.log('history', this.state.discuss.history.messages)
```
Comment (alexkuhn) [inline addons/mail/static/src/new/core/messaging.js]: I removed this `console.log`

File: addons/mail/static/src/new/core/messaging.js [hunk 2]
```diff
@@ -353,13 +353,14 @@ export class Messaging {
                     break;
                 }
                 case "mail.message/mark_as_read": {
-                    for (const message_id of notif.payload.message_ids) {
+                    const {message_ids: messageIds, needaction_inbox_counter} = notif.payload;
```
Comment (alexkuhn) [inline addons/mail/static/src/new/core/messaging.js]: prettier unhappy with missing whitespace inside `{}`

File: addons/mail/static/src/new/thread/message.js [hunk 3]
```diff
@@ -153,10 +153,21 @@ export class Message extends Component {
...
+        }
+        return this.messaging.state.threads[this.props.threadLocalId].id === "inbox"; 
+    }
+
     toggleStar() {
         this.messaging.toggleStar(this.props.message.id);
     }
 
+    setDone() {
+        messaging.setDone(this.props.message.id);
```
Comment (alexkuhn) [inline addons/mail/static/src/new/thread/message.js]: Missing `this` to access `messaging`.

Somehow, this actually works. It leaks some technical details on context in an owl template, in which the variable `messaging` is available.

File: addons/mail/static/src/new/thread/message.js [hunk 4]
```diff
@@ -153,10 +153,21 @@ export class Message extends Component {
...
+    }
```
Comment (alexkuhn) [inline addons/mail/static/src/new/thread/message.js]: Can just call `messaging.setDone` immediately in template

File: addons/mail/static/tests/new/discuss/discuss_tests.js [hunk 5]
```diff
@@ -884,3 +884,201 @@ QUnit.test("basic top bar rendering", async function (assert) {
...
+        " on Refactoring"
+    );
+    assert.strictEqual(
+        message.querySelectorAll(".o-mail-message-actions .cursor-pointer").length,
+        4
+    );
+    assert.strictEqual(message.querySelectorAll(".o-mail-message-action-add-reaction").length, 1);
+    assert.strictEqual(message.querySelectorAll(".o-mail-message-action-toggle-star").length, 1);
+    assert.strictEqual(message.querySelectorAll(".o-mail-message-action-reply-to").length, 1);
+    assert.strictEqual(message.querySelectorAll(".o-mail-message-action-mark-read").length, 1);
```
Comment (alexkuhn) [inline addons/mail/static/tests/new/discuss/discuss_tests.js]: `queryAll.length` can be `contains` assertions

File: addons/mail/static/tests/new/discuss/discuss_tests.js [hunk 6]
```diff
@@ -884,3 +884,201 @@ QUnit.test("basic top bar rendering", async function (assert) {
...
+    await openDiscuss();
+    assert.strictEqual(target.querySelectorAll(".o-mail-message").length, 1);
+    const message = target.querySelector(".o-mail-message");
+    assert.strictEqual(message.querySelectorAll(".o-mail-message-recod-name").length, 1);
+    assert.strictEqual(
+        message.querySelector(".o-mail-message-recod-name").textContent,
+        " on Refactoring"
+    );
+    assert.strictEqual(
+        message.querySelectorAll(".o-mail-message-actions .cursor-pointer").length,
```
Comment (alexkuhn) [inline addons/mail/static/tests/new/discuss/discuss_tests.js]: `cursor-pointer` is pure style. The intend here is more on `i` (`i` meaning icon)

File: addons/mail/static/tests/new/discuss/discuss_tests.js [hunk 7]
```diff
@@ -884,3 +884,201 @@ QUnit.test("basic top bar rendering", async function (assert) {
...
+        target.querySelector('button[data-mailbox="history"]').classList.contains("o-active")
+    );
+    assert.strictEqual(
+        target.querySelectorAll('.o-mail-discuss-content .o-mail-thread [data-empty-thread=""]')
+            .length,
+        1
+    );
+
+    await click('button[data-mailbox="inbox"]');
+    assert.ok(target.querySelector('button[data-mailbox="inbox"]').classList.contains("o-active"));
```
Comment (alexkuhn) [inline addons/mail/static/tests/new/discuss/discuss_tests.js]: `assert.hasClass` is best that truthy `classList.contains`

File: addons/mail/static/tests/new/discuss/discuss_tests.js [hunk 8]
```diff
@@ -884,3 +884,201 @@ QUnit.test("basic top bar rendering", async function (assert) {
...
+            .length,
+        1
+    );
+
+    await click('button[data-mailbox="history"]');
+    assert.ok(
+        target.querySelector('button[data-mailbox="history"]').classList.contains("o-active")
+    );
+    assert.strictEqual(
+        target.querySelectorAll(`.o-mail-discuss-content .o-mail-thread [data-empty-thread=""]`)
```
Comment (alexkuhn) [inline addons/mail/static/tests/new/discuss/discuss_tests.js]: Avoid backquote string if you can. They are slightly annoying to work with :P

File: addons/mail/static/tests/new/discuss/discuss_tests.js [hunk 9]
```diff
@@ -884,3 +884,201 @@ QUnit.test("basic top bar rendering", async function (assert) {
...
+    );
+    assert.strictEqual(
+        target.querySelectorAll(".o-mail-discuss-content .o-mail-thread .o-mail-message").length,
+        2
+    );
+
+    await click('.o-mail-discuss-actions button[data-action="mark-all-read"]');
+    assert.ok(target.querySelector('button[data-mailbox="inbox"]').classList.contains("o-active"));
+    assert.strictEqual(
+        target.querySelectorAll('.o-mail-discuss-content .o-mail-thread [data-empty-thread=""]')
```
Comment (alexkuhn) [inline addons/mail/static/tests/new/discuss/discuss_tests.js]: I slightly prefer `$(target).find()` over `target.querySelector()`, because the former has less token for the technical stuffs, thus the meaningful part is more apparent.

File: addons/mail/static/tests/new/discuss/discuss_tests.js [hunk 10]
```diff
@@ -884,3 +884,201 @@ QUnit.test("basic top bar rendering", async function (assert) {
...
+    pyEnv["mail.notification"].create([
+        {
+            mail_message_id: mailMessageId1, // id of related message
+            notification_type: "inbox",
+            res_partner_id: pyEnv.currentPartnerId, // must be for current partner
+        },
+        {
+            mail_message_id: mailMessageId2, // id of related message
+            notification_type: "inbox",
+            res_partner_id: pyEnv.currentPartnerId, // must be for current partner
```
Comment (alexkuhn) [inline addons/mail/static/tests/new/discuss/discuss_tests.js]: Could remove the comment (too noisy)

**Review by alexkuhn — COMMENTED** [review]
All introduced tests are actually old tests.
Should also remove the old tests, otherwise we don't see duplicate.

**Comment by alexkuhn** [comment]
👆I just applied the comments
